tests/test_request_media.py

```diff
@@ -0,0 +1,119 @@
+import pytest
+
+import falcon
+from falcon import errors, media, testing
+
+
+def create_client(handlers=None):
+    res = testing.SimpleTestResource()
+
+    app = falcon.API()
+    app.add_route('/', res)
+
+    if handlers:
+        app.req_options.media_handlers.update(handlers)
+
+    client = testing.TestClient(app)
+    client.resource = res
+
+    return client
+
+
+@pytest.mark.parametrize('media_type', [
+    (None),
+    ('*/*'),
+    ('application/json'),
+    ('application/json; charset=utf-8'),
+])
+def test_json(media_type):
+    client = create_client()
+    expected_body = b'{"something": true}'
+    headers = {'Content-Type': media_type}
+    client.simulate_post('/', body=expected_body, headers=headers)
+
+    media = client.resource.captured_req.media
+    assert media is not None
+    assert media.get('something') is True
+
+
+@pytest.mark.parametrize('media_type', [
+    ('application/msgpack'),
+    ('application/msgpack; charset=utf-8'),
+    ('application/x-msgpack'),
+])
+def test_msgpack(media_type):
+    client = create_client({
+        'application/msgpack': media.MessagePackHandler(),
+        'application/x-msgpack': media.MessagePackHandler(),
+    })
+    expected_body = b'\x81\xc4\tsomething\xc3'
```

Comment:
[fix] Test an object that contains both `unicode` (encoded to UTF-8), and `bytes` (encoded to the MessagePack bin type), to ensure the media handler is correctly decoding and distinguishing the two types.